Added min/max metadata to numerics and a default binning strategy [WIP]

senior opened the pull request:
This commit adds new min/max metadata to all numeric columns. That
metadata is used in the automatic binning of breakout fields. This
commit also extends MBQL to indicate that a breakout column (or
columns) should be binned using a "default" strategy. The query must
also include the maximum number of bins. Using the number of bins with
the min/max values, the bin width is computed and used in the SQL
query to group data.

-------------------------------------------

This approach to automatic binning is probably the easiest but it does have some drawbacks. The query includes a "max number of bins". If there are no values for a bin, it does not emit a row. This is a by-product of how SQL works. It seems like this might be unintuitive for users. If the user said they wanted 20 bins, my guess is they'd probably like to see the begin/end range of the bin and 0 for the number of entries for the given bin. The breakout_test changes included in this PR show that unintuitive behavior. The test asks for 20 bins, but the data is very clumpy, so all of the data fits in 5 bins. Creating 20 bins with 0s in the empty bins would be a significantly more complicated query for the query processor, but it should be possible.

Something that @salsakran and I have discussed are various ways that users would want to bin discrete values (number of bins, user defined bins, regex on text) in addition to various defaults or predefined binning strategy values. This approach here assumes that the query includes the information needed to execute the query. With this approach, we could still have an admin setup and name an "age range bin" or "version regex" however the querying for that default/predefined bin would happen earlier in the query pipeline. That strategy would be expanded earlier in the pipeline and query processor would still be able to translate the query and execute it straight away. The query would look something like:

```
{:breakout [[:binning-strategy :version :user "version prefix regex"]]}
```

Which would expand internally to:

```
{:breakout [[:binning-strategy :version :regex "^(v.*)/.*" 1]]}
```

We wouldn't need to expose the flexibility to the user via the UI, it could be just an internal detail which allows a clear separation between query construction and user defined settings. It does seem like allowing users to supply this might be a nice future feature though. The user is clicking into a breakout, uses the admin defined regex, they want something a little bit different. The back-end would just get it's regex from the user rather than a setting.

iethree committed:
Added min/max metadata to numerics and a default binning strategy

This commit adds new min/max metadata to all numeric columns. That
metadata is used in the automatic binning of breakout fields. This
commit also extends MBQL to indicate that a breakout column (or
columns) should be binned using a "default" strategy. The query must
also include the maximum number of bins. Using the number of bins with
the min/max values, the bin width is computed and used in the SQL
query to group data.

senior commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -158,17 +176,18 @@
+                                     (as (formatted field) field)))
+    (apply h/group new-hsql (map (fn [field]
+                                   (if (instance? BinnedField field)
+                                     (qualified-alias field)
```
This is ugly and probably not what we want. I'm not sure if this is done elsewhere, but what I have is a computed projected value that is aliased (via the `as` above) that I want to reference in the group by and the order by. Using formatted in each place (as the code was before) it will repeat that computation in the select, group by and order by. With the above change, the query looks like:

```
SELECT count(*) AS "count",
       (floor(("PUBLIC"."VENUES"."LATITUDE" / 2.0)) * 2.0) AS "LATITUDE",
       (floor(("PUBLIC"."VENUES"."LONGITUDE" / 5.0)) * 5.0) AS "LONGITUDE"
FROM "PUBLIC"."VENUES"
GROUP BY "LATITUDE", "LONGITUDE"
ORDER BY "LATITUDE", "LONGITUDE"
```

senior commented on src/metabase/query_processor/expand.clj:
```diff
@@ -181,6 +181,10 @@
 
 ;;; ## breakout & fields
 
+(s/defn ^:ql ^:always-validate binning-strategy :- FieldPlaceholder
```
The strategy-name isn't used here. Right now I just have the automatic binning strategy so I didn't use it. My next steps here would be using that to figure out what kind of binning the user wanted. Values here would be regex, automatic, predefined etc.

senior commented on src/metabase/query_processor/interface.clj:
```diff
@@ -128,6 +130,11 @@
   (getName [_] (name field)))
 
+(s/defrecord BinnedField [field :- Field
+                          num-bins  :- s/Int]
```
This will need to change to include the strategy and the parameter (i.e. regex, set of bins etc). Not sure if that will just all be loaded in this BinnedField, or if that equates to different types.

tlrobinson commented:
Maybe you have already discussed these things, but I had a few questions:

1. Is specifying the number of bins always preferable to specifying the bin size directly (which would be significantly simpler to implement, I think)?
  * Our datetime binning is essentially by size (month, day, year), not number of bins, for example.
  * Obviously the downside to specifying bin size it's easy to select a range + bin size with too many or too few bins
  * Maybe it would be sufficient to guess the initial bin size and ensure subsequent actions (e.x. drill through, "brush" filtering #4978) update the bin size to something sensible?
2. The min/max of the table doesn't necessarily correspond with the min/max of the query. If you have a > or < filter on the binned field itself you can guess a better min/max, but if you have a filter on a field that's correlated with the binned field (say for some reason you're binning "age" and filtering "date of birth", which are are separate fields) then the bin size could be significantly off. Maybe that's ok?

senior commented:
Good questions @tlrobinson. I think for question 1, I think the answer is no. And it's more apparent with this PR than in the previous PR. The previous one had a globally defined number of bins. Assuming we don't know the distribution of values and the user doesn't either, having a set number of bins seems the most useful to me. I would imagine that if a user understands the distributions better (or if we could via @thearthur's work) we could maybe put together a better value there. One of the difficult things with this stuff is depending on how the data is skewed, it can look pretty bad. Maybe the back end supports bin width OR bin size ultimately to account for this.

Question 2 is interesting. I think what I'd want as a user in that case is to be able to specify a different bin size (just for this query) or a different number of bins. The SQL query this would generate would need to be more complex too. We'd not want to use the min/max we stored, we'd want to get the min/max with the filter/criteria applied so we can adjust our bins accordingly.

senior commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -81,6 +88,17 @@
 
+  BinnedField
+  (formatted [{{:keys [min-value max-value] :as field} :field bins :num-bins}]
+    (let [bin-width (Math/ceil (/ (- (Math/floor max-value)
```
BTW, putting this logic in the protocol extensions was a nice refactor @tlrobinson. I stole it from an earlier version of https://github.com/metabase/metabase/pull/4700

iethree committed:
Number of bins now optional

This adds a new global setting to change the default (currently
8). When the number of bins isn't provided in the query, the default
it used.

senior commented:
@tlrobinson @camsaul The current plan, after talking with @salsakran is to work on getting this merged in and add the other binning strategies, including adding the ability to supply bin width, as a "stage 2" of this feature. Currently this only returns the bins with values (not empty bins) that will also be fixed as part of a separate PR/task.

iethree committed:
Fix breakout test failure on MySQL

The issue only shows up in Circle CI and appears to be related to
MySQL there returning java.lang.Float values instead of
doubles. Converting these to doubles (via round-to-decimals) before
comparing fixes the issue.

tlrobinson commented:
Random thought: it would be nice if we did the same min/max thing for date fields. The frontend could use it to guess an appropriate date breakout `unit` (year, month, day), e.x. if when breaking out you were to click the date column name without selecting a unit, or when you've zoomed in to hours/minutes and clear the date filter, and suddenly are viewing 3 years of data broken out by minute.

tlrobinson commented:
If sync hasn't populated min/max then running a query throws `java.lang.NullPointerException`. We should make sure the frontend doesn't allow you to use the number of bins binning strategy and/or handle that gracefully in the QP.

senior commented:
Closing this and reopening as https://github.com/metabase/metabase/pull/5331
